### ISSUE TITLE
Invalid format-security warning with empty argpack

### ISSUE THREAD

**Issue by msharov** [llvm/llvm-project#56583]
```c++
#include <stdio.h>

template <typename... Args>
void print_error (const char* fmt, Args... args)
{
    printf (fmt, args...);
}

int main (void)
{
    print_error ("message\n");
    return 0;
}
```
Produces incorrect `-Wformat-security` error, even though `fmt` is the string literal `"message\n"`:
```console
tes.cc:6:13: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
    printf (fmt, args...);
            ^~~
tes.cc:11:5: note: in instantiation of function template specialization 'print_error<>' requested here
    print_error ("message\n");
    ^
tes.cc:6:13: note: treat the string as an argument to avoid this
    printf (fmt, args...);
            ^
            "%s", 
1 warning generated.
```
No warning is emitted when `print_error` is given additional arguments.

**Comment by inclyc** [comment]
https://godbolt.org/z/cnvhE59hE

**Comment by msharov** [comment]
Now wait just a minute here. A passthrough printf wrapper is a very common and perfectly acceptable use case. Is it really a good policy to force everyone to use a pragma to shut up this warning? This is especially problematic because `-Wformat-security` is on by default and in most cases we do want these warnings. Here clang can't see that `fmt` is a literal, and is intended as a passthrough format string. This is NOT a security problem, and should not get this invalid warning.

**Comment by inclyc** [comment]
Sorry for closing this, did you mean that `clang` shouldn't give such a warning if the argument is a literal for all cases where this template function is used?

**Comment by inclyc** [comment]
as you can see `const char* fmt` is obviously non-literal

**Comment by inclyc** [comment]
Did you expect that the "literal" property need to be propagated? Similar to the fact that it is a literal in this function call, we should think that this procedure should not generate a warning? If a non-literal value is used in another place, then we fire a warning for that.

See https://godbolt.org/z/hb5vjx9GY

**Comment by inclyc** [comment]
In my opinion, literal properties should not be propagated into functions (i.e. from caller to callee). `constexpr` ,`consteval` ... these expressions, which are evaluated at **compile-time**, should be handled correctly (which seems to fail at https://github.com/llvm/llvm-project/issues/55805 however).

But `const char* fmt` cannot be considered a "literal", even though the argument is a literal. You may be able to make a workaround based on explicitly mentioned **compile-time** constants, by keywords such as `constexpr`. 

## Note 1:
It is possible that, in another `.cpp` file invokes `print_error`: 

```C++
// OK.cpp
template <typename... Args>
void print_error (const char* fmt, Args... args)
{
    printf (fmt, args...);
}
```

```C++
// buggy.cpp
print_error("%d%d%d"); // may cause overflow
print_error(some_user_input_variable);  // may cause overflow
```

and if these linked together, which is fairly dangerous and may cause overflow. And this is actually the security issue warned by clang.

## Note 2:
`const char* fmt` does not mean "fmt" is a **compile-time** constant. It just means you can not change its value in function body.

**Comment by msharov** [comment]
The issue here is merely that the compiler is giving a warning about code that is usually not a security problem. Yes, it can occasionaly be one, as your examples illustrate, but not in the sort of code where this warning usually is given for. Printf wrappers are used all over the place for things like debug logging or error reporting, and are never going to be a security problem. Either the format string is a literal, or it is chosen from some set of hardcoded values. No unsanitized user input is ever going to end up there. The compiler, on the other hand, freaks out and assumes the worst anyway.

This paranoia may well be desirable for some lint-type checking, but not for an on-by-default compiler option. format-security has no business just assuming by default that unsanitized user input will be used as the format string. More information is needed to make the decision; information that the compiler is never going to get.

Now, there are a couple of ways to fix this. One would be to propagate the literal status when the template is inlined. Another would be to allow setting a printf args attribute on the template and disabling the warning when the arg is explicitly marked as a format string. That would require adding a separate template prototype somehow? I'm not sure what that syntax looks like.

And finally, you could split the format-security warning in two. One would check the args validity, and the other would check for the nonliteral format string. The latter should then be off by default, because there are many legitimate uses for nonliteral format strings that produce no security issues and the compiler should not assume it knows better unless it is asked to do so.

**Comment by msharov** [comment]
I tried adding the format attribute:
```c++
template <typename... Args>
__attribute__((__format__(__printf__,1,2))) void print_error (const char* fmt, Args... args)
{
    printf (fmt, args...);
}
```
And this generates:
```
tes.cc:4:50: error: format attribute requires variadic function
__attribute__((__format__(__printf__,1,2))) void print_error (const char* fmt, Args... args)
```
Should I submit this as a separate issue? For this purpose, an argpack ought to be treated as a variadic arg. Just how hard would it be to support that?

**Comment by inclyc** [comment]
> The issue here is merely that the compiler is giving a warning about code that is usually not a security problem. Yes, it can occasionaly be one, as your examples illustrate, but not in the sort of code where this warning usually is given for. Printf wrappers are used all over the place for things like debug logging or error reporting, and are never going to be a security problem. Either the format string is a literal, or it is chosen from some set of hardcoded values. No unsanitized user input is ever going to end up there. The compiler, on the other hand, freaks out and assumes the worst anyway.
> 
> This paranoia may well be desirable for some lint-type checking, but not for an on-by-default compiler option. format-security has no business just assuming by default that unsanitized user input will be used as the format string. More information is needed to make the decision; information that the compiler is never going to get.
> 
> Now, there are a couple of ways to fix this. One would be to propagate the literal status when the template is inlined. Another would be to allow setting a printf args attribute on the template and disabling the warning when the arg is explicitly marked as a format string. That would require adding a separate template prototype somehow? I'm not sure what that syntax looks like.
> 
> And finally, you could split the format-security warning in two. One would check the args validity, and the other would check for the nonliteral format string. The latter should then be off by default, because there are many legitimate uses for nonliteral format strings that produce no security issues and the compiler should not assume it knows better unless it is asked to do so.

Well, if you don't like this warning, just turn it off by `-Wno-format-nonliteral` `-Wno-format-security` 

There is a certain difficulty in splitting this warning into two as you said. This is because the compiler must make sure that fmt can be statically evaluated at compile time before it can check whether the arguments correspond. 

For your example(i.e. printf wrappers), we can not determine whether the template function argument `const char *fmt` is a **literal**. Although your codes may just use hard-coded or literal only, there may also a potential call with unsanitized argument (e.g user input), since clang cannot see other translation units performing sematic checking. If you can ensure that every references to this function call is definitely safe, just disable this warning by -Wno-format-security`. 
 
The problem may be that we don't have a good way to make `constexpr`, `consteval` these explicitly marked `fmt` checks that the compiler can determine the value of. The crux of this question is **compile-time evaluation**. `const char* fmt` cannot be evaluated at **compile-time** so currently clang gives you this warning. If you declare that template function `static`, it is somehow necessary to consider `fmt` as literal because we have known all used places of that function, and not linked by some other.

**Comment by inclyc** [comment]
> I tried adding the format attribute:
> 
> ```c++
> template <typename... Args>
> __attribute__((__format__(__printf__,1,2))) void print_error (const char* fmt, Args... args)
> {
>     printf (fmt, args...);
> }
> ```
> 
> And this generates:
> 
> ```
> tes.cc:4:50: error: format attribute requires variadic function
> __attribute__((__format__(__printf__,1,2))) void print_error (const char* fmt, Args... args)
> ```
> 
> Should I submit this as a separate issue? For this purpose, an argpack ought to be treated as a variadic arg. Just how hard would it be to support that?

This is NOT a bug, template functions instantiate from every used cases, these functions are apparently not variadic.

## Example

### Souce

```C++
template <typename... Args> 
void print_error(const char *fmt, 
Args... args) {
  printf(fmt, args...);
}
```
and call this template function as:

```C++
print_error("abcd");
```

### Compile

the compiler instantiate this template function for you, using Abstract-Syntax-Tree Transform. looks like:

```C++
// What the compiler see when performing sematic format string checking
void print_error(const char *fmt) {
  printf(fmt);
}
```

This is not a variadic function.

**Comment by msharov** [comment]
> Well, if you don't like this warning, just turn it off by
> -Wno-format-nonliteral -Wno-format-security

It isn't that I dislike the warning, it is that you go too far in applying it. Just like with many other warnings added over the last decade, I get to spend a lot of time and effort figuring out "the approved way" of writing every particular idiom. Not because the way I'm doing it is wrong, but because you arbitrarily decided that it is somehow more bug-prone and so should be disallowed. How many real bugs in my code did these warnings found? Zero. All I got was a lot of unnecessary labor. I don't claim that my code is perfect; only that I never get the type of bug you work so hard to warn about.

Yes, it's your compiler, and your rules, so I'm going to let this go.
I just can't help being somewhat annoyed.
